src/coreclr/tools/Common/TypeSystem/Common/TypeDesc.cs

```diff
@@ -633,6 +633,17 @@ public bool HasFinalizer
             }
         }
 
+        /// <summary>
+        /// Gets a value indicating the finalizer associated with this class is critical.
+        /// </summary>
+        public bool HasCriticalFinalizer
```

Comment:
Critical finalizers are not a core type system concept. We do not need them to introduce in managed type system. We can just check for the critical finalizer base type in ComputeFlagsEx when building EEType.

Reply:
Right. This should be very similar to `HasEagerFinalizerFlag`, just a different trait. `HasCriticalFinalizer` should be computed at the same time when `HasEagerFinalizerFlag` is computed, no need to compute it earlier.

A few things to note:
- `CriticalFinalizerObject` does not need to be immediate base, so you will need to get all the way up the inheritance chain until you see `CriticalFinalizerObject` or nothing.
- you do not need to do that to every type, it would make sense to first check if it is a finalizable type (typedesc would know that).

Reply:
I reworked this so the logic is only in `EETypeBuilderHelpers`. Is this what you had in mind?